Added phase for two phase evolution of pinned grain size.

alarshi opened the pull request:
This PR adds the necessary functions, phase distribution function and computation of mean grain size from roughness, for grain size evolution.

Commit:
Added phase for two phase evolution of pinned grain size.

Commit:
Added phase for two phase evolution of pinned grain size.

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -1064,6 +1109,12 @@ namespace aspect
                              "When set to zero, grain size will not be reduced. "
                              "Units: \\si{\\meter}.");
+          prm.declare_entry ("Volume fraction of phase", "0.4",
+                             Patterns::List (Patterns::Double (0., 1.)),
```
Why is this a list of doubles rather than just a double? From the description, it sounds like it would be just one number.

alarshi replied:
You are right, it should be only double since we decided on using only one phase with the pinned state formulation.

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -1064,6 +1109,12 @@ namespace aspect
+          prm.declare_entry ("Volume fraction of phase", "0.4",
+                             Patterns::List (Patterns::Double (0., 1.)),
+                             "The volume fraction of one of the phases in the two phases evolved using Bercovici and "
+                             "Richard (2012). The volume fraction of the other phase can be simply calculated by subtracting "
```
```suggestion
                             "Ricard (2012). The volume fraction of the other phase can be simply calculated by subtracting "
```
I think this might also be confusing, since everywhere else we mention the Mulyokova paper. So if the user is not aware that they use this damage model, they might not know that this parameter is only used when we use the pinned damage model. I think this would be worth adding.

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -1313,6 +1364,14 @@ namespace aspect
+          volume_fraction_phase_one             = Utilities::string_to_double
+                                                  (Utilities::split_string_list(prm.get ("Volume fraction of phase")));
+
+          if (volume_fraction_phase_one.size()>1)
+            for (unsigned int i=0; i<volume_fraction_phase_one.size(); ++i)
+              AssertThrow( (volume_fraction_phase_one[i] == 0. || volume_fraction_phase_one[i] == 1.),
+                           ExcMessage("Error: Volume fraction should be between (0, 1) to use two phase flow !"));
```
I am confused by this code. What is it supposed to do? Did you test that this works as intended? 
To me, it looks like now you can only pick 0 or 1 as volume fractions...

I think what you want to do might be this:
```suggestion
          if (volume_fraction_phase_one.size()>1)
            for (unsigned int i=0; i<volume_fraction_phase_one.size(); ++i)
              AssertThrow( (volume_fraction_phase_one[i] != 0. && volume_fraction_phase_one[i] != 1.),
                           ExcMessage("Error: Volume fraction must be between (0, 1) to use two phase damage!"));
```

alarshi replied:
Yes, this is what I wanted to do. Thank you!

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -1064,6 +1109,12 @@ namespace aspect
                              "Units: \\si{\\meter}.");
+          prm.declare_entry ("Volume fraction of phase", "0.4",
+                             Patterns::List (Patterns::Double (0., 1.)),
+                             "The volume fraction of one of the phases in the two phases evolved using Bercovici and "
```
```suggestion
                             "The volume fraction of one of the phases in the two-phase damage model of Bercovici and "
```

jdannberg reviewed (commented):
I didn't get to look through everything; I'll add more comments later.

Commit:
fix typos.

Commit:
Use it as a reference.

alarshi commented:
Thanks! I addressed some of your comments in the latest commits and also added a jupyter notebook that includes the relevant text for the equations in computing grain size from roughness. I still have to modify the equations accordingly.

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -211,6 +211,51 @@ namespace aspect
+    GrainSize<dim>::
+    moment_of_grain_size_distribution (const int n) const
+    {
+      // This function normalizes the grain size distribution using the nth moment.
+      // Description can be found in eq 8 of Bercovici and Ricard (2012)
```
This type of description should be in the .h file.

jdannberg replied:
It would also be good to explain a bit more why we need to do this. Because the use in the function below does not look like it normalizes the grain size (i.e. it gets an unsigned int, not a grain size).

jdannberg commented on include/aspect/material_model/grain_size.h:
```diff
@@ -275,6 +275,23 @@ namespace aspect
+         */
+        std::vector<double> volume_fraction_phase_one;
+
+        double moment_of_grain_size_distribution (const int n) const;
```
Can this be an unsigned int or can it be negative?
```suggestion
        double moment_of_grain_size_distribution (const unsigned int n) const;
```

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -211,6 +211,51 @@ namespace aspect
+    {
+      // This function normalizes the grain size distribution using the nth moment.
+      // Description can be found in eq 8 of Bercovici and Ricard (2012)
+      const double sigma = 0.8;
```
Can you explain what this variable does (or why it is 0.8)?

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -211,6 +211,51 @@ namespace aspect
+      // This factor is used to convert the roughness equation into mean grain size
+      // Refer to Appendix H.1, eqs 8, F.28 in Bercovici and Ricard (2012) for more details.
+      const double b1 = 1./20 ;
+      const double c1 = 3 * b1 * moment_of_grain_size_distribution(4) / (8 * moment_of_grain_size_distribution (2));
```
```suggestion
      const double c1 = 3.0 * b1 * moment_of_grain_size_distribution(4) / (8.0 * moment_of_grain_size_distribution (2));
```

Also, if this is the only place where it is used, I wonder if it is really worth making it a  separate function. We can discuss in our meeting today.

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -211,6 +211,51 @@ namespace aspect
+    {
+      // This factor is used to convert the roughness equation into mean grain size
+      // Refer to Appendix H.1, eqs 8, F.28 in Bercovici and Ricard (2012) for more details.
+      const double b1 = 1./20 ;
```
```suggestion
      const double b1 = 1./20.;
```

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -211,6 +211,51 @@ namespace aspect
+      const double h1 = c1 * (1 - volume_fraction_phase_one);
+      const double h2 = c1 * (1 - volume_fraction_phase_two);
+
+      return (volume_fraction_phase_one /sqrt (h1)  + volume_fraction_phase_two/sqrt(h2)) ;
```
```suggestion
      return (volume_fraction_phase_one / std::sqrt(h1)  + volume_fraction_phase_two / std::sqrt(h2)) ;
```

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -1345,6 +1396,14 @@ namespace aspect
                  ExcMessage("The parameter 'Use paleowattmeter' has been removed. "
                             "Use the parameter 'Grain size evolution formulation instead'."));
 
+          volume_fraction_phase_one             = Utilities::string_to_double
+                                                  (Utilities::split_string_list(prm.get ("Volume fraction of phase")));
```
This is just a single double, not a list of doubles, right?

alarshi replied:
Yes, thank you!

jdannberg commented on include/aspect/material_model/grain_size.h:
```diff
@@ -275,6 +275,23 @@ namespace aspect
+         * Mechanisms for the generation of plate tectonics by two-phase grain-damage
+         * and pinning. Physics of the Earth and Planetary Interiors 202 (2012): 27-55.
+         */
+        std::vector<double> volume_fraction_phase_one;
```
Why is this a vector of doubles rather than just a double?

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -1345,6 +1396,14 @@ namespace aspect
+          volume_fraction_phase_one             = Utilities::string_to_double
+                                                  (Utilities::split_string_list(prm.get ("Volume fraction of phase")));
+
+          if (volume_fraction_phase_one.size()>1)
```
I think we want this to be a double anyway; but even with a vector, you would still need to check that the volume fraction is between zero and one if it only had one entry.

jdannberg reviewed (commented):
I have commented on the changes you made, but it's hard for me to tell if this is what is needed exactly because this PR only adds the parameters, but does not use them in the computation of material properties. I think it would be good to add that as well (we can talk about it in the meeting today).

jdannberg commented on contrib/phase_distribution_function.ipynb:
```diff
@@ -0,0 +1,183 @@
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXMAAAEECAYAAADandTrAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAIABJREFUeJzt3Xl81PW97/HXJ8mE7AmQEEhCQGUXlLBJ1bbS45FeN6g9WvDg7W2P19NW7G3POZ7Wtte9tbf22J5q2+vS5dgqtLYsirborbUCLcqSAIrsewIkLIFskO17/5hJZAkwyczkN8v7+XjMY8hvfjPzyfdB3vnm+/3+vmPOOUREJLYleV2AiIiETmE ...
+      "text/plain": [
```
Could you remove the output?

jdannberg commented on contrib/phase_distribution_function.ipynb:
```diff
@@ -0,0 +1,183 @@
+{
```
Can you add a line that says what the file does and move it to [contrib/grain_size/phase_distribution_function.ipynb](https://github.com/grain-damage-Mulyukova/aspect/pull/5/files#diff-dc46a07843680930bcb01e7514aef17cebbb27d6154ea9c8d07810abe5cc8aed)

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -211,6 +211,51 @@ namespace aspect
+    template <int dim>
+    double
+    GrainSize<dim>::
+    moment_of_grain_size_distribution (const int n) const
```
@gassmoeller says to make this into an anonymous function by deleting it from the header file and to hide it by putting it into an anontmous name space. 

```suggestion
namespace
  {
    double
    moment_of_grain_size_distribution (const int n) const
    [...]
   }
```

Commit:
Merge conflicts.

Commit:
Simply the notebook.

Commit:
Add implementation of the functions.

alarshi commented:
Thank you for the review! I added the implementation of these functions in the latest commit. I still need to test the implementation.

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -270,9 +319,21 @@ namespace aspect
-          const double grain_size_growth_rate = grain_growth_rate_constant[phase_index] / (m * std::pow(grain_size,m-1))
-                                                * std::exp(- (grain_growth_activation_energy[phase_index] + pressure * grain_growth_activation_volume[phase_index])
-                                                           / (constants::gas_constant * temperature));
+          double grain_size_growth_rate = 0.0;
+
+          // grain growth rate in the two-phase damage model
+          if (grain_size_evolution_formulation == Formulation::pinned_grain_damage)
+            grain_size_growth_rate = grain_growth_rate_constant[phase_index] * geometric_constant[phase_index] * phase_distribution_function (volume_fraction_phase_one)
+                                     / (m * std::pow(grain_size,m-1) * std::pow(roughness_to_grain_size_factor(volume_fraction_phase_one), m))
+                                     * std::exp(- (grain_growth_activation_energy[phase_index] + pressure * grain_growth_activation_volume[phase_index])
+                                                / (constants::gas_constant * temperature));
+          else
+            grain_size_growth_rate = grain_growth_rate_constant[phase_index] / (m * std::pow(grain_size,m-1))
+                                     * std::exp(- (grain_growth_activation_energy[phase_index] + pressure * grain_growth_activation_volume[phase_index])
+                                                / (constants::gas_constant * temperature));
```
This looks correct to me, but I would implement it in fewer lines:

```suggestion
double grain_size_growth_rate = grain_growth_rate_constant[phase_index] / (m * std::pow(grain_size,m-1))
* std::exp(- (grain_growth_activation_energy[phase_index] + pressure * grain_growth_activation_volume[phase_index])
/ (constants::gas_constant * temperature));
                                                           
// in the two-phase damage model grain growth depends on the proportion of the two phases                         
if (grain_size_evolution_formulation == Formulation::pinned_grain_damage)
  grain_size_growth_rate *= geometric_constant[phase_index] * phase_distribution_function (volume_fraction_phase_one) / 
std::pow(roughness_to_grain_size_factor(volume_fraction_phase_one), m)
```
That also makes it easier to see what the differences are.

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -211,6 +211,55 @@ namespace aspect
+    roughness_to_grain_size_factor (const double volume_fraction_phase_one)  const
+    {
+      // This factor is used to convert the roughness equation into mean grain size
+      // Refer to Appendix H.1, eqs 8, F.28 in Bercovici and Ricard (2012) for more details.
```
I think this should go into the .h file.

jdannberg reviewed (approved):
Looks good to me. I just have two small suggestions, but we should definitely go ahead and use this for our tests!

jdannberg commented on source/material_model/grain_size.cc:
```diff
@@ -211,6 +211,55 @@ namespace aspect
+      double
+      moment_of_grain_size_distribution (const unsigned int n)
+      {
+        const double sigma = 0.8;
```
Can you add a comment saying that this is the variance of the log-normal distribution?

jdannberg commented on contrib/grain_size/phase_distribution_function.ipynb:
```diff
@@ -0,0 +1,145 @@
+{
```
I also just went through all equations for our discussion with Elvira, and I would definitely keep the notebook to explain where the equations are coming from.

alarshi replied:
Got it, thanks!

Commit:
Addressed comments.

alarshi commented:
Thank you for the review! I addressed your comments in the latest commit. I also tested that the implementation is consistent with our interpretation of the equations: test case results in PR#13.
